src/eduid/satosa/scimapi/scim_attributes.py

```diff
@@ -67,27 +82,106 @@ def process(
+                data_owner = _sorted_scopes[0]
+                logger.debug(f'Using first scope as data owner: {_sorted_scopes}')
         if not data_owner:
+            logger.warning(f'No data owner for issuer {data.auth_info.issuer}')
             return None
-        userdb = self._userdbs.get(data_owner)
-        if not userdb:
-            logger.error(f'Found no userdb for data owner {data_owner}')
-            return None
+
+        logger.info(f'entityId {data.auth_info.issuer}, scope(s) {scopes}, data_owner {data_owner}')
+
+        userdb = self.get_userdb_for_data_owner(data_owner)
+
         _ext_ids = data.attributes.get(self.ext_id_attr, [])
-        if _ext_ids:
-            ext_id = _ext_ids[0]
-            user = userdb.get_user_by_external_id(ext_id)
-            if user:
-                logger.info(
-                    f'Found SCIM user {user.scim_id} using {self.ext_id_attr} {ext_id} (data owner: {data_owner})'
-                )
-            else:
-                logger.info(f'No user found using {self.ext_id_attr} {ext_id}')
-            return user
-        return None
+        if len(_ext_ids) != 1:
+            logger.warning(f'Got more or less than one externalId using attribute {self.ext_id_attr}: {_ext_ids}')
+            return None
+
+        ext_id = _ext_ids[0]
+        user = userdb.get_user_by_external_id(ext_id)
+        if user:
+            logger.info(f'Found SCIM user {user.scim_id} using {self.ext_id_attr} {ext_id} (data owner: {data_owner})')
+        else:
+            logger.info(f'No user found using {self.ext_id_attr} {ext_id}')
+        return user
+
+
+def _extract_saml_scope(idpsso: List[Mapping[str, Any]]) -> Set[str]:
+    """
+    Extract scopes from SAML extension data that looks something like this:
+
+    [
+    {
+        "__class__": "urn:oasis:names:tc:SAML:2.0:metadata&IDPSSODescriptor",
+        "extensions": {
+            "__class__": "urn:oasis:names:tc:SAML:2.0:metadata&Extensions",
+            "extension_elements": [
+                {
+                    "__class__": "urn:mace:shibboleth:metadata:1.0&Scope",
+                    "text": "eduid.se",
+                    "regexp": "false"
+                },
+                {
+                    "__class__": "urn:oasis:names:tc:SAML:metadata:ui&UIInfo",
+                    ...
+                }
+            ]
+        }
+    }]
+    """
+    res = set()
+    for top_level in idpsso:
+        if top_level.get('__class__') != 'urn:oasis:names:tc:SAML:2.0:metadata&IDPSSODescriptor':
+            continue
+        if 'extensions' in top_level:
+            if top_level['extensions'].get('__class__') == 'urn:oasis:names:tc:SAML:2.0:metadata&Extensions':
+                for element in top_level['extensions'].get('extension_elements', []):
+                    if element.get('__class__') != 'urn:mace:shibboleth:metadata:1.0&Scope':
+                        continue
+                    _text = element.get('text')
```

Comment:
This is working but it is semi-correct. The `shibmd:Scope` element is allowed to be a regex in which  case we need to do regex-match to deduct if the scope is correct.

So, here, we should store that attribute and the part that checks if the scope is known

```python
                if _scope in self.config.scope_to_data_owner:
                    data_owner = self.config.scope_to_data_owner[_scope]
```

should be converted to match the regex.. (which makes this ugly).

Reference: https://wiki.shibboleth.net/confluence/display/SC/ShibMetaExt+V1.0

PS: I am not sure how common it is to define Scopes as regex to justify doing this work, but in any case, it's good to know about it.